crates/flashblocks-rpc/src/state.rs

```diff
@@ -60,44 +59,55 @@ where
 {
     pub fn new(client: Client) -> Self {
         Self {
-            pending_block: Arc::new(ArcSwapOption::new(None)),
+            pending_blocks: Arc::new(ArcSwapOption::new(None)),
             flashblock_sender: broadcast::channel(BUFFER_SIZE).0,
             metrics: Metrics::default(),
             client,
         }
     }
 
     pub fn on_canonical_block_received(&self, block: &RecoveredBlock<OpBlock>) {
-        if let Some(cur) = self.pending_block.load_full() {
-            if cur.block_number() <= block.number {
-                // clear the pending flashblockblock
-                if let Some(prev) = self.pending_block.swap(None) {
+        if let Some(pending_blocks) = self.pending_blocks.load_full() {
+            if pending_blocks.latest_block_number() <= block.number {
+                if let Some(prev) = self.pending_blocks.swap(None) {
                     self.metrics.pending_clear_catchup.increment(1);
                     self.metrics
                         .pending_snapshot_height
-                        .set(prev.block_number() as f64);
+                        .set(prev.latest_block_number() as f64);
                     self.metrics
                         .pending_snapshot_fb_index
-                        .set(prev.flashblock_index() as f64);
+                        .set(prev.latest_flashblock_index() as f64);
                 }
+            } else {
+                // Clear everything not necessary
+                // and re-process flashblocks
+                let mut flashblocks = pending_blocks.get_flashblocks();
+                flashblocks.retain(|flashblock| flashblock.metadata.block_number > block.number);
```

Comment:
Can do in a follow up, but I think we should check if there's a flashblock/canonical reorg here and if so clear everything (as the flashblocks are built on bad state)

Reply:
wait nvm this wont work since block hash will change with presence of state root